ivadomed/config/config_pose.json

```diff
@@ -0,0 +1,94 @@
+{
+    "command": "train",
+    "gpu_ids": [0],
+    "path_output": "data/example/labeling_t0test1",
+    "model_name": "pose_model",
+    "debugging": false,
+    "loader_parameters": {
+        "path_data": ["data/data-multi-subject"],
+        "target_suffix": ["_heatmap0"],
+        "extensions": [".nii.gz"],
+        "roi_params": {
+            "suffix": null,
+            "slice_filter_roi": null
+        },
+        "contrast_params": {
+            "training_validation": ["T1w"],
+            "testing": [ "T1w"],
+            "balance": {}
+        },
+        "slice_filter_params": {
+            "filter_empty_mask": true,
+            "filter_empty_input": true
+        },
+        "slice_axis": "sagittal",
+        "multichannel": false,
+	    "soft_gt": true
+    },
+    "split_dataset": {
+        "fname_split": null,
+        "random_seed": 8,
+        "split_method" : "participant_id",
+        "data_testing": {"data_type": null, "data_value":[]},
+        "balance": null,
+        "train_fraction": 0.6,
+        "test_fraction": 0.2
+    },
+    "training_parameters": {
+        "batch_size": 1,
+        "loss": {
+            "name": "JointsMSELoss"
+        },
+        "training_time": {
+            "num_epochs": 100,
+            "early_stopping_patience": 100,
+            "early_stopping_epsilon": 0.001
+        },
+        "scheduler": {
+            "initial_lr":0.0005,
+            "lr_scheduler": {
+                "name": "CosineAnnealingLR",
+                "base_lr": 1e-6,
+                "max_lr": 1e-4
+            }
+        },
+        "balance_samples": false,
+        "mixup_alpha": null,
+        "transfer_learning": null  
+    },
+    "default_model": {
+        "name": "HourglassNet"
+    },
+    "FiLMedUnet": {
+        "applied": false,
+        "metadata": "contrasts",
+        "film_layers": [0, 1, 0, 0, 0, 0, 0, 0, 0, 0]
+    },
+    "uncertainty": {
+        "epistemic": false,
+        "aleatoric": false,
+        "n_it": 0
+    },
+    "postprocessing": {
+        "remove_noise": {"thr": -1},
+        "binarize_prediction": {"thr": 0.5},
+        "uncertainty": {"thr": -1, "suffix": "_unc-vox.nii.gz"},
+        "remove_small": {"unit": "vox", "thr": 3}
+    },
+    "evaluation_parameters": {
+        "target_size": {"unit": "vox", "thr": [20, 100]},
+        "overlap": {"unit": "vox", "thr": 3}
+    },
+    "transformation": {
+      "Resample":
+        {
+            "wspace": 256,
+            "hspace": 256,
+            "dspace": 1,
+            "flag_pixel": true
+        },
+      "VertebralSplitting": {"max_joint": 11, "applied_to": ["gt"]},
+      "NumpyToTensor": {},
+      "NormalizeInstance": {"applied_to": ["im"]}
+    }
```

Comment:
## Comparing HourglassNet (Reza) to Countception (Lucas)

I'm currently trying to figure out how to use `vertebral-labeling-validation` to compare the two approaches, however I've run into an issue:

* Lucas' model was trained on [straightened data](https://github.com/ivadomed/ivadomed/blob/b15d7cfd4b271261835b4c93864d92411c45c3d6/ivadomed/config/config_vertebral_labeling.json#L16-L17) (because `sct_label_vertebrae` uses straightened data for template-matching with the PAM50 template).
* Reza's model, however, is currently trained on [unstraightened data](https://github.com/ivadomed/ivadomed/blob/a2929e689ad5a3a55a66e39f2c54f98af36b6cff/ivadomed/config/config_pose.json#L16-L17) from `data-multi-subject`.

This makes Reza's model much trickier to integrate into the [current structure of `sct_label_vertebrae`](https://github.com/spinalcordtoolbox/spinalcordtoolbox/pull/2679/files#r710454872), which in turn makes it much trickier to evaluate Reza's approach using the `vertebral-labeling-validation` repo.

To fix this, I'm thinking that training the HourglassNet with straightened data would make our lives much easier down the line when trying to evaluate the approach +  integrate it into SCT.

Reply:
> To fix this, I'm thinking that training the HourglassNet with straightened data would make our lives much easier down the line when trying to evaluate the approach + integrate it into SCT

good idea.

Reply:
Small progress update: I'm still currently working on getting [`vertebral-labeling-validation`](https://github.com/sct-pipeline/vertebral-labeling-validation) set up and running. There are [some existing issues](https://github.com/sct-pipeline/vertebral-labeling-validation/issues) with the scripts, but I'm trying to sort those out right now, and I will open a PR in that repo when I have a clear set of working instructions for running the test procedure.

Reply:
Small update 2: I have a model trained on straightened data now!

![sub-hamburg05__rec-mid_T1w__0](https://user-images.githubusercontent.com/16181459/135386122-216f36ca-db9f-4030-92ca-1bb37f47bdd2.gif)

(Arriving here took a little longer than I hoped, because the precise preprocessing commands used for the Countception model weren't documented beyond a [short paragraph description](https://github.com/neuropoly/vertebral-labeling-deep-learning#preprocessing). This time, I'll make sure to share a script that will help with reproduction.)

But, now that the pieces are in place, I should be able to now work towards comparing the two models using [`vertebral-labeling-validation`](https://github.com/sct-pipeline/vertebral-labeling-validation).

Reply:
This is fantastic @joshuacwnewton ! few comments:
- the image quantization seems low-- i am just wondering if this is caused by the GIF generation, or if the low quality was used during training
- the straightened cord seems squeezed in the superior-inferior direction. I am wondering if this is just something that happened during the GIF creation or also before, during training.
- did you see changes in performance between the straightened and the non-straightened scenario?
- can you please remind me again what was the rationale for doing the inference in the straightened scenario: was it to:
  - be more compatible with the current workflow of sct_label_vertebrae
  - allow for easier comparison between the current method and the method you are implementing
  - supposedly better performance on the straightened cord because less variability (and cropping around spinal cord made easier)?
- (minor comment: the head/feet is upside down)

Reply:
> the image quantization seems low-- i am just wondering if this is caused by the GIF generation, or if the low quality was used during training

I believe the low quantization is caused by the GIF generation. It is present in Reza's original gif, too: https://github.com/ivadomed/ivadomed/issues/766#issuecomment-835714961

<details>
<summary> Here is the (higher-quality) input image corresponding to the gif </summary>

![Screenshot from 2021-09-30 12-07-33](https://user-images.githubusercontent.com/16181459/135491626-43ad2022-d933-4fe8-95e9-11fee38169f2.png)
</details>

> the straightened cord seems squeezed in the superior-inferior direction. I am wondering if this is just something that happened during the GIF creation or also before, during training.

The SI squeezing occurs during training. This is because the model configuration currently resamples images to 256x256, but `sct_straightened_spinalcord` crops images around the spinal cord, so the rectangular image is reshaped to a square image.

The resampling is present because I had been re-using the config file from Reza's initial PR as-is, just to get things working. 

> did you see changes in performance between the straightened and the non-straightened scenario?

This is something I want to evaluate more thoroughly, and so I don't want to comment prematurely, but I'll get back to you on this. :slightly_smiling_face: 

> can you please remind me again what was the rationale for doing the inference in the straightened scenario: was it to: 

Initially, the motivation was for the first two reasons that you describe (`sct_label_vertebrae` compatibility, ease of comparison). 

Specifically, working with curved data would require a much bigger refactor on SCT's end. But, working on straightened data means that we could swap the new model very easily into the [existing SCT PR](https://github.com/spinalcordtoolbox/spinalcordtoolbox/pull/2679), since it would work with the same input data as Lucas' model.

However, after I learned that `sct_straighted_spinalcord` includes cropping around the spinal cord, I began to think about the other reasons you describe -- i.e. that straightening would reduce variability between subjects, and cropping would reduce the # of non-SC pixels and thus reduce the class imbalance further. So, I agree that these are secondary motivations to preprocess the data in this way.

> (minor comment: the head/feet is upside down)

I believe this is due to differences in the location of the origin between numpy arrays and NIfTI images (top-left vs bottom-left). When I open the images in FSLeyes, they're correctly oriented. So, I was thinking that it might be an issue with the gif functionality in ivadomed.

Still, if it's a visual bug with `--gif`, I feel like it should have been noticed already... So, maybe it _is_ a problem on my end. (Insight from the ivadomed team would be appreciated, e.g. if others have used the `--gif` functionality without this issue.)

Reply:
> This is because the model configuration currently resamples images to 256x256, but sct_straighten_spinalcord crops images around the spinal cord (resulting in a rectangular image), so it gets squished during resampling to the square dimensions.

I see. That might be a red flag, as we want to conserve the image proportion during the resampling. I remember there was a conversation about that between Reza, Charley and me (probably buried somewhere in the PR). But long story short: what we want is an algorithm that _resamples_ to a 256x256 square (ie: keeping the physical proportion of the image in case the input image is not isotropic resolution), versus _resizing_ (ie: re-defining the physical size of each pixel). If I'm not clear, I can explain better in a meeting

> Still, if it's a visual bug with --gif, I feel like it should have been noticed already... 

Not necessarily, because so far ivadomed has mostly been used with axial images, where swapping top/bottom doesn't look "weird".

Reply:
> > This is because the model configuration currently resamples images to 256x256, but sct_straighten_spinalcord crops images around the spinal cord (resulting in a rectangular image), so it gets squished during resampling to the square dimensions.
> 
> I see. That might be a red flag, as we want to conserve the image proportion during the resampling. I remember there was a conversation about that between Reza, Charley and me (probably buried somewhere in the PR). But long story short: what we want is an algorithm that _resamples_ to a 256x256 square (ie: keeping the physical proportion of the image in case the input image is not isotropic resolution), versus _resizing_ (ie: re-defining the physical size of each pixel). If I'm not clear, I can explain better in a meeting

Ahhh, I see! Thank you for bringing this up again, because it actually turns out that in [the thread you've referenced](https://github.com/ivadomed/ivadomed/pull/787#discussion_r645318419) containing the discussion from Charley, the ultimate conclusion was [this](https://github.com/ivadomed/ivadomed/pull/787#discussion_r645318419):

> Ok, I will update the config file with `Center Crop` instead of `resize`.

It looks like this was never completed in Reza's PR, and I missed it as well, so the resizing was still being applied. I'll make that change ASAP.